state-chain/ethereum-eip712/src/eip712.rs

```diff
@@ -0,0 +1,1032 @@
+			domain: EIP712Domain,
+			types: Types,
+			#[serde(rename = "primaryType")]
+			primary_type: String,
+			message: scale_value::Value,
+		}
+
+		#[derive(Deserialize)]
+		#[serde(untagged)]
+		enum Type {
+			Val(TypedDataHelper),
+			String(String),
+		}
+
+		match Type::deserialize(deserializer)? {
+			Type::Val(v) => {
+				let TypedDataHelper { domain, types, primary_type, message } = v;
+				Ok(TypedData { domain, types, primary_type, message })
+			},
+			Type::String(s) => {
+				let TypedDataHelper { domain, types, primary_type, message } =
+					serde_json::from_str(&s).map_err(serde::de::Error::custom)?;
+				Ok(TypedData { domain, types, primary_type, message })
+			},
+		}
+	}
+}
+
+// === impl TypedData ===
+
+impl Eip712 for TypedData {
+	type Error = Eip712Error;
+
+	fn domain(&self) -> Result<EIP712Domain, Self::Error> {
+		Ok(self.domain.clone())
+	}
+
+	fn type_hash() -> Result<[u8; 32], Self::Error> {
+		Err(Eip712Error::Message("dynamic type".to_string()))
+	}
+
+	fn struct_hash(&self) -> Result<[u8; 32], Self::Error> {
+		let tokens = encode_data(&self.primary_type, &self.message.clone(), &self.types)?;
+		Ok(keccak256(encode(&tokens)))
+	}
+
+	/// Hash a typed message according to EIP-712. The returned message starts with the EIP-712
+	/// prefix, which is "1901", followed by the hash of the domain separator, then the data (if
+	/// any). The result is hashed again and returned.
+	fn encode_eip712(&self) -> Result<[u8; 32], Self::Error> {
+		let domain_separator = self.domain.separator();
+		let mut digest_input = [&[0x19, 0x01], &domain_separator[..]].concat().to_vec();
+
+		if self.primary_type != "EIP712Domain" {
+			// compatibility with <https://github.com/MetaMask/eth-sig-util>
+			digest_input.extend(&self.struct_hash()?[..])
+		}
+		Ok(keccak256(digest_input))
+	}
+}
+
+/// Represents the name and type pair
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct Eip712DomainType {
+	pub name: String,
+	#[serde(rename = "type")]
+	pub r#type: String,
+}
+
+/// Encodes an object by encoding and concatenating each of its members.
+///
+/// The encoding of a struct instance is `enc(value₁) ‖ enc(value₂) ‖ … ‖ enc(valueₙ)`, i.e. the
+/// concatenation of the encoded member values in the order that they appear in the type. Each
+/// encoded member value is exactly 32-byte long.
+///
+///   - `primaryType`: The root type.
+///   - `data`: The object to encode.
+///   - `types`: Type definitions for all types included in the message.
+///
+/// Returns an encoded representation of an object
+pub fn encode_data(
+	primary_type: &str,
+	data: &scale_value::Value,
+	types: &Types,
+) -> Result<Vec<Token>, Eip712Error> {
+	let hash = hash_type(primary_type, types)?;
+	let mut tokens = vec![Token::Uint(U256::from(hash))];
+
+	if let Some(fields) = types.get(primary_type) {
+		for (field, value) in
+			fields.iter().zip(data.get_scale_value_fields().map_err(|e| {
```

Comment:
Are we guaranteed the the fields in the Eip712DomainType are in the same order as the fields in the `scale_value::Value`?